src/storage/data_table.cpp

```diff
@@ -318,4 +318,53 @@ void DataTable::DeallocateVarlensOnShutdown(RawBlock *block) {
     }
   }
 }
+
+bool DataTable::HasConflict(const transaction::TransactionContext &txn, const TupleSlot slot) const {
+  UndoRecord *const version_ptr = AtomicallyReadVersionPtr(slot, accessor_);
+  return HasConflict(txn, version_ptr);
+}
+
+bool DataTable::IsVisible(const transaction::TransactionContext &txn, const TupleSlot slot) const {
+  UndoRecord *version_ptr;
+  bool visible;
+  do {
+    version_ptr = AtomicallyReadVersionPtr(slot, accessor_);
+    // Here we will need to check that the version pointer did not change during our read. If it did, the content
+    // we have read might have been rolled back and an abort has already unlinked the associated undo-record,
+    // we will have to loop around to avoid a dirty read.
+    // TODO(Matt): might not need to read visible in the loop (move after?) but not confident without large random tests
+    visible = Visible(slot, accessor_);
+  } while (version_ptr != AtomicallyReadVersionPtr(slot, accessor_));
+
+  // Nullptr in version chain means no version visible to any transaction alive at this point.
+  // Alternatively, if the current transaction holds the write lock, it should be able to read its own updates.
+  if (version_ptr == nullptr || version_ptr->Timestamp().load() == txn.TxnId().load()) {
+    return visible;
+  }
+
+  // Apply deltas until we determine a version safe for us to read
+  // If the version chain becomes null, this tuple does not exist for this version, and the last delta
+  // record would be an undo for insert that sets the primary key to null, which is intended behavior.
```

Comment:
I don't think this comment is accurate anymore.  With GC in place we cannot infer anything about visibility simply by the absence of undo records and the last delta is not guaranteed to be an insert.

The comment should also be updated at line 225 as well.